File: examples/java/ToolServer.java
package examples.java;

import uniffi.smpp_codec_ffi.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.List;
import java.util.ArrayList;

public class ToolServer {
    private static final int PORT = 2775;
    private static final int HEADER_LEN = 16;

    // Command IDs
    private static final int CMD_BIND_RECEIVER = 0x00000001;
    private static final int CMD_BIND_TRANSMITTER = 0x00000002;
    private static final int CMD_SUBMIT_SM = 0x00000004;
    private static final int CMD_UNBIND = 0x00000006;
    private static final int CMD_BIND_TRANSCEIVER = 0x00000009;

    public static void main(String[] args) {
        ExecutorService threadPool = Executors.newCachedThreadPool();
        
        try (ServerSocket server = new ServerSocket(PORT)) {
            System.out.println("Java Multi-threaded Server listening on port " + PORT);

            while (true) {
                Socket client = server.accept();
                System.out.println("Client connected: " + client.getInetAddress());
                threadPool.submit(() -> handleClient(client));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void handleClient(Socket client) {
        try (InputStream input = client.getInputStream();
             OutputStream output = client.getOutputStream()) {

            while (true) {
                byte[] header = new byte[HEADER_LEN];
                int read = input.read(header);
                if (read < HEADER_LEN) break;

                ByteBuffer buffer = ByteBuffer.wrap(header);
                int commandLen = buffer.getInt();
                int commandId = buffer.getInt();
                int commandStatus = buffer.getInt();
                int sequenceNumber = buffer.getInt();

                System.out.printf("Header: Len=%d, ID=0x%08X, Seq=%d%n", commandLen, commandId, sequenceNumber);

                int bodyLen = commandLen - HEADER_LEN;
                byte[] body = new byte[bodyLen];
                if (bodyLen > 0) {
                    input.read(body);
                }

                byte[] fullPdu = new byte[commandLen];
                System.arraycopy(header, 0, fullPdu, 0, HEADER_LEN);
                System.arraycopy(body, 0, fullPdu, HEADER_LEN, bodyLen);

                List<UByte> fullPduList = convertToUByteList(fullPdu);
                List<UByte> responsePdu = null;

                if (commandId == CMD_BIND_RECEIVER || commandId == CMD_BIND_TRANSMITTER || commandId == CMD_BIND_TRANSCEIVER) {
                    BindRequest req = SmppCodecFfiKt.decodeBindRequest(fullPduList);
                    System.out.println("Received Bind: " + req.getSystemId());
                    BindResponse resp = new BindResponse(
                        sequenceNumber, commandId | 0x80000000, req.getSystemId(), null
                    );
                    responsePdu = SmppCodecFfiKt.encodeBindResponse(resp);
                } else if (commandId == CMD_SUBMIT_SM) {
                    SubmitSmRequest req = SmppCodecFfiKt.decodeSubmitSmRequest(fullPduList);
                    System.out.println("Received SubmitSm");
                    
                    // Concatenation Detection (Example)
                    boolean isUdh = (req.getEsmClass() & 0x40) == 0x40;
                    if (isUdh) System.out.println("  [DETECTION] UDH Detected");
                    
                    SubmitSmResponse resp = new SubmitSmResponse(sequenceNumber, "MsgID_12345");
                    responsePdu = SmppCodecFfiKt.encodeSubmitSmResponse(resp);
                } else if (commandId == CMD_UNBIND) {
                    System.out.println("Received Unbind");
                    UnbindResponse resp = new UnbindResponse(sequenceNumber);
                    responsePdu = SmppCodecFfiKt.encodeUnbindResponse(resp);
                }

                if (responsePdu != null) {
                    byte[] respBytes = new byte[responsePdu.size()];
                    for (int i = 0; i < responsePdu.size(); i++) {
                        respBytes[i] = (byte)responsePdu.get(i).getValue();
                    }
                    output.write(respBytes);
                    System.out.println("Sent Response");
                    if (commandId == CMD_UNBIND) break;
                }
            }
        } catch (Exception e) {
            System.err.println("Error handling client: " + e.getMessage());
        } finally {
            try { client.close(); } catch (Exception ignored) {}
        }
    }

    private static List<UByte> convertToUByteList(byte[] bytes) {
        List<UByte> list = new ArrayList<>();
        for (byte b : bytes) {
            list.add(new UByte((short)(b & 0xFF)));
        }
        return list;
    }
}

File: examples/kotlin/ToolServer.kt
import uniffi.smpp_codec_ffi.*
import java.net.ServerSocket
import java.net.Socket
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import kotlin.concurrent.thread

const val HOST = "127.0.0.1"
const val PORT = 2775
const val HEADER_LEN = 16

// Command IDs
const val CMD_BIND_RECEIVER = 0x00000001.toInt()
const val CMD_BIND_TRANSMITTER = 0x00000002.toInt()
const val CMD_QUERY_SM = 0x00000003.toInt()
const val CMD_SUBMIT_SM = 0x00000004.toInt()
const val CMD_DELIVER_SM = 0x00000005.toInt()
const val CMD_UNBIND = 0x00000006.toInt()
const val CMD_REPLACE_SM = 0x00000007.toInt()
const val CMD_CANCEL_SM = 0x00000008.toInt()
const val CMD_BIND_TRANSCEIVER = 0x00000009.toInt()
const val CMD_ENQUIRE_LINK = 0x00000015.toInt()
const val CMD_SUBMIT_MULTI_SM = 0x00000021.toInt()
const val CMD_ALERT_NOTIFICATION = 0x00000102.toInt()
const val CMD_DATA_SM = 0x00000103.toInt()
const val CMD_BROADCAST_SM = 0x00000111.toInt()
const val CMD_QUERY_BROADCAST_SM = 0x00000112.toInt()
const val CMD_CANCEL_BROADCAST_SM = 0x00000113.toInt()

fun main() {
    val server = ServerSocket(PORT)
    println("Server listening on port $PORT")

    while (true) {
        val client = server.accept()
        println("Client connected: ${client.inetAddress}")
        thread {
            handleClient(client)
        }
    }
}

fun handleClient(client: Socket) {
    try {
        val input = client.getInputStream()
        val output = client.getOutputStream()

        while (true) {
            val header = ByteArray(HEADER_LEN)
            if (input.read(header) < HEADER_LEN) {
                break
            }

            val buffer = ByteBuffer.wrap(header)
            val commandLen = buffer.getInt()
            val commandId = buffer.getInt()
            val commandStatus = buffer.getInt()
            val sequenceNumber = buffer.getInt()

            println(String.format("Header: Len=%d, ID=0x%08X, Status=%d, Seq=%d", commandLen, commandId, commandStatus, sequenceNumber))

            val bodyLen = commandLen - HEADER_LEN
            val body = ByteArray(bodyLen)
            if (bodyLen > 0) {
                input.read(body)
            }

            val fullPdu = header + body
            // Note: uniffi expects List<Byte> (or UByte).
            // We assume List<Byte> for compatibility with encodeToByteArray().toList()
            val fullPduList = fullPdu.toList() 

            var responsePdu: List<Byte>? = null

            when (commandId) {
                CMD_BIND_RECEIVER, CMD_BIND_TRANSMITTER, CMD_BIND_TRANSCEIVER -> {
                    val req = decodeBindRequest(fullPduList)
                    println("Received Bind: ${req.systemId}")
                    val resp = BindResponse(
                        sequenceNumber = sequenceNumber.toUInt(),
                        commandId = (commandId or -0x80000000).toUInt(), // 0x80000000 as int is negative
                        systemId = req.systemId,
                        scInterfaceVersionTlv = null
                    )
                    responsePdu = encodeBindResponse(resp)
                }
                CMD_SUBMIT_SM -> {
                    val req = decodeSubmitSmRequest(fullPduList)
                    println("Received SubmitSm: ${req.sourceAddr} -> ${req.destinationAddr}")
                    
                    val isUdh = (req.esmClass.toInt() and 0x40) == 0x40
                    var sarRef: Int? = null
                    var sarTotal: Int? = null
                    var sarSeq: Int? = null
                    
                    for (tlv in req.tlvs) {
                        when (tlv.tag.toInt()) {
                            Tags.SAR_MSG_REF_NUM.toInt() -> {
                                // Extract u16 from value (2 bytes, big endian)
                                if (tlv.value.size >= 2) {
                                    sarRef = ((tlv.value[0].toInt() and 0xFF) shl 8) or (tlv.value[1].toInt() and 0xFF)
                                }
                            }
                            Tags.SAR_TOTAL_SEGMENTS.toInt() -> {
                                if (tlv.value.isNotEmpty()) {
                                    sarTotal = tlv.value[0].toInt() and 0xFF
                                }
                            }
                            Tags.SAR_SEGMENT_SEQNUM.toInt() -> {
                                if (tlv.value.isNotEmpty()) {
                                    sarSeq = tlv.value[0].toInt() and 0xFF
                                }
                            }
                        }
                    }
                    
                    if (isUdh) {
                        println("  [DETECTION] UDH Concatenation detected")
                    }
                    if (sarRef != null) {
                        println("  [DETECTION] SAR Concatenation detected: Ref=$sarRef, Part=$sarSeq/$sarTotal")
                    }

                    val resp = SubmitSmResponse(
                        sequenceNumber = sequenceNumber.toUInt(),
                        messageId = "MsgID_12345"
                    )
                    responsePdu = encodeSubmitSmResponse(resp)
                }
                CMD_UNBIND -> {
                    println("Received Unbind")
                    val resp = UnbindResponse(sequenceNumber = sequenceNumber.toUInt())
                    responsePdu = encodeUnbindResponse(resp)
                }
                // Add other cases as needed...
                else -> {
                    println(String.format("Unknown/Unhandled Command ID: 0x%08X", commandId))
                }
            }

            if (responsePdu != null) {
                val respBytes = ByteArray(responsePdu.size)
                for (i in responsePdu.indices) {
                    respBytes[i] = responsePdu[i]
                }
                output.write(respBytes)
                println("Sent Response")
                
                if (commandId == CMD_UNBIND) {
                    break
                }
            }
        }
        client.close()
    } catch (e: Exception) {
        println("Error handling client: $e")
        e.printStackTrace()
    }
}
